### PR TITLE
fix interval units parsing

### LINKED ISSUE #11271: `interval` cast has some problematic behaviour
### Describe the bug

There's a few issues with casting to an `interval`.

Specifically:
* `'1'::interval` -> `0 years 1 mons 0 days 0 hours 0 mins 0.000000000 secs` - postgres returns `1s`, I think other databases do the same - this is the worst offender IMHO
* `'1s'::interval` gives an error, with postgres it returns `1s`
* `'1second'::interval` gives an error, with postgres it returns `1s`

Happy to try to fix this if there's willingness? In particular, I think `'1'::interval` should be fixed asap.

### To Reproduce

_No response_

### Expected behavior

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by samuelcolvin**
## Which issue does this PR close?

Fixes #11271 (accidentally got closed the the companion PR https://github.com/apache/arrow-rs/pull/6211).

Negative PR (at least by lines of Rust) 🎉 .

## Rationale for this change

Use the much improved unit parsing logic now supported by arrow-rs, instead of a simple 'add "seconds"' hack

## What changes are included in this PR?

Change interval expression parsing to use arrow-rs logic

## Are these changes tested?

Yes

## Are there any user-facing changes?

Improvements to `interval` expression functionality.

### CODE REVIEW

File: datafusion/sqllogictest/test_files/interval.slt [hunk 1]
```diff
@@ -560,6 +585,22 @@ select '1 month'::interval + ts from t;
...
+5 days 1 hours
+
+# trailing extra unit, this matches PostgreSQL
+query ?
+select interval '5 day 0' hour
+----
+5 days
+
+# This is interpreted as "0 hours" with PostgreSQL, should be fixed with
+query error DataFusion error: Arrow error: Parser error: Invalid input syntax for type interval: "5 day HOUR"
```
Comment (samuelcolvin) [inline datafusion/sqllogictest/test_files/interval.slt:601]: see https://github.com/apache/arrow-rs/issues/6390

**Comment by samuelcolvin** [comment]
@alamb if this passes I believe it's ready to merge despite the very minor regression, overall it's way better than what went before.

In particular expressions like `interval '1s'` are pretty common and without this, fail with a very ugly error:

```
Arrow error: Parser error: Invalid input syntax for type interval: "1s seconds"
```

**Review by alamb — APPROVED** [review]
Thank you @samuelcolvin  -- this looks like a great improvement to me. 🙏

**Comment by alamb** [comment]
This PR has a merge conflict now and it turns out there seems to have been a logical conflict with a few of the prior PRs that were merge. I will fix them in this PR


Screenshot 2024-09-15 at 8 05 45 AM

**Comment by alamb** [comment]
Thanks again @samuelcolvin

**Pull request merged**